### PR TITLE
Biophysical models

### PR DESCRIPTION
**Pull request by karalets**
Adding a biophysical likelihood to our regression suite so we can regress models according to the concentrations etc, but actually regress deltaG under the hood.

This is the initial PR, I do not expect tp merge that until we have working examples, so let's leave it up here for a bit.

Design principles:
This model currently lives in regressors.
One needs to pass a regressor (i.e. GP, NN) into this model which is the base_regressor and is used to model deltaG.
On top of that, a deterministic transformation is applied to samples from that regressor to transform the latent deltaG into a measurement. Finally, we have homoschedastic noise on top of the measurement.

This model should work fine with either GP or NN regressors as base_regressors and should be usale anywhere where we currently use regressors in our codebase as the interface is consistent.

@yuanqing-wang  and @miretchin you guys can have a look.

### CODE REVIEW

File: pinot/regressors/biophysical_regressor.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+# =============================================================================
+# IMPORTS
+# =============================================================================
+import torch
+import pinot
+import abc
+import math
+from pinot.regressors.base_regressor import BaseRegressor
+from pinot.regressors.gaussian_process_regressor import ExactGaussianProcessRegressor
+from pinot.regressors.neural_network_regressor import NeuralNetworkRegressor
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:10]: both of these imports are unused

File: pinot/regressors/biophysical_regressor.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+        sigma_m = torch.exp(self.log_sigma_measurement)
+
+        distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
+
+        return distribution_measurement
+
+
+    def loss(self, h=None, y=None, test_ligand_concentration=1e-3, **kwargs):
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:54]: note irregular tabs

File: pinot/regressors/biophysical_regressor.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    """
+
+    def __init__(self, base_regressor=None):
+        super(BiophysicalRegressor, self).__init__()
+        self.base_regressor = base_regressor
+        self.log_sigma_measurement = torch.zeros(1)
+
+
+	def g(self, func_value=None, test_ligand_concentration=1e3-):
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:35]: test_ligand_concentration here is invalid syntax

File: pinot/regressors/biophysical_regressor.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	    return 1 / (1 + torch.exp(-func_value) / test_ligand_concentration)
+
+
+	def condition(self, h=None, test_ligand_concentration=1e-3, **kwargs):
+        distribution_base_regressor = self.base_regressor.condition(h, kwargs)
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:40]: double check everytime you use kwargs to use **kwargs

File: pinot/regressors/biophysical_regressor.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        #we sample from the latent f to paush things through the likelihood
+        #Note: if we do this, in order to get good estimates of LLK we may need to draw multiple samples
+        f_sample = distribution_base_regressor.rsample()
+
+        mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+        sigma_m = torch.exp(self.log_sigma_measurement)
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:47]: sigma is always 1? I don't see where log_sigma_measurement is ever updated after it's initialized as 0.

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:47]: it will e a parameter next time you see it

File: pinot/regressors/biophysical_regressor.py (hunk 4 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:40]: yup fixed all those, will push in a it

File: pinot/regressors/biophysical_regressor.py (hunk 3 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:35]: indeed!

File: pinot/regressors/biophysical_regressor.py (hunk 2 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:54]: gosh darn

File: pinot/regressors/biophysical_regressor.py (hunk 1 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:10]: yeah I guess my design ideas about this class changed as I was doing it. Lint would catch all these issues, but has not run yet.

File: pinot/regressors/biophysical_regressor.py (hunk 2 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py]: better?

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py]: yeah I mean I prefer 4 spaces to tabs, but at least now it's consistent 🙂

File: pinot/regressors/biophysical_regressor.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    """
+
+    def __init__(self, base_regressor=None):
+        super(BiophysicalRegressor, self).__init__()
+        self.base_regressor = base_regressor
+        self.log_sigma_measurement = torch.nn.Parameter(torch.zeros(1))
+
+
+	def g(self, func_value=None, test_ligand_concentration=1e-3):
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:30]: I feel like these should be a positional parameters. do you have other plans for it?

File: pinot/regressors/biophysical_regressor.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+
+    def marginal_loss(self, h=None, y=None, test_ligand_concentration=1e-3, n_samples=10, **kwargs):
+    	"""
+    	sample n_samples often from loss in order to get a better approximation
+    	"""
+    	distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+    	marginal_loss_measurement = 0
+    	for ns in range(n_samples):
+    		for ns in range(n_samples):
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:70]: this seems strange to me. perhaps it could have two different params, one for the inner vs. outer loop?

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:70]: elaborate please

File: pinot/regressors/biophysical_regressor.py (hunk 6 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:30]: elaborate ... I don't know what you mean.

File: pinot/regressors/biophysical_regressor.py (hunk 2 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py]: let's pick a linter style and enforce that on all the code. This should not be down to choice.
If we pick e.g.  flake8 , we can stick to it and be consistent.

File: pinot/regressors/biophysical_regressor.py (hunk 6 above)
Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:30]: just that I think of keyword arguments as denoting something optional, but these [func_value, test_ligand_concentration] are not optional

File: pinot/regressors/biophysical_regressor.py (hunk 7 above)
Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:70]: one is for sampling from the base_regressor posterior (inner loop), and the other is for sampling from the loss function (outer loop). in this case, we've given them always the same number of samples for each, but it doesn't have to be that way, right?

actually I'm really confused now because I see that you're just setting f_sample, so the inner loop isn't doing anything; you're just using the results from the most recent loop.

maybe I'm misunderstanding what this is doing.

File: pinot/regressors/biophysical_regressor.py (hunk 6 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:30]: I see, it's a good distinction.

File: pinot/regressors/biophysical_regressor.py (hunk 7 above)
Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:70]: similar question applies for the for loop inside `marginal_sample` actually (lines 54-55)

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:70]: I am sampling both each time... I don;t see the need for an inner and an outer loop I just go:
```
dG ~ N(mu_regressor, sigma_regressor)
m ~ N(g(dG,c), sigma_measurement)
```

and voila, that is my generative model.

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:70]: If you look closely, for the marginal loss I am actually trying to marginalize out the deltaG, that might correspond to your inner loop idea.

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:70]: that makes more sense to me, but there are nested loops inside that function, not a single loop

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:70]: I'm just asking what the function of these lines are, found in both marginal_loss and marginal_sample
```
   		for ns in range(n_samples):
   			f_sample = distribution_base_regressor.rsample()
```

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py]: > that makes more sense to me, but there are nested loops inside this function (as currently written), not a single loop

Yeah, that's what's known as a copy-paste bug. I commited a change, good catch.

File: pinot/regressors/biophysical_regressor.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+   	def marginal_sample(self, h=None, n_samples=100, test_ligand_concentration=1e-3, **kwargs):
+   		distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+   		samples_measurement = []
+   		for ns in range(n_samples):
+   			f_sample = distribution_base_regressor.rsample()
+	        mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+	        sigma_m = torch.exp(self.log_sigma_measurement)
+	        distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
+    		samples_measurement.append(distribution_measurement.sample())
```
Comment (miretchin) [inline pinot/regressors/biophysical_regressor.py:59]: ok the source of my confusion is that these lines need to be in the for-loop

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:59]: technically ```sigma_m = torch.exp(self.log_sigma_measurement)``` does not need to be in there, but the others do. I just left it there to group the parameters for the measurement distribution.

The reason is: the mean is being re-calculated every time we sample a new deltaG

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:59]: confusion resolved?

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:59]: no, these need to be moved a tab forward, inside the for-loop. that's what I'm referring to 😀

otherwise you'll end up with len(samples_measurement) == 1 and it contains the distribution_measurement.sample() corresponding to mu_m from just one f_sample

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:59]: but fair enough, you could do something like:

```
   	        sigma_m = torch.exp(self.log_sigma_measurement)
		samples_measurement = []
   		for ns in range(n_samples):
   			f_sample = distribution_base_regressor.rsample()
	        	mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
	        	distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
    			samples_measurement.append(distribution_measurement.sample())
```

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:59]: I agree it's not as nice-looking though

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:59]: wait, in my code they ARE inside the for loop...

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:59]: Screen Shot 2020-07-01 at 10 17 54 PM

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:59]: what I see:

image

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:59]: probably another weird tab thing..

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py:59]: ok, that explains our differences. Let me check this.

Comment (miretchin) [reply pinot/regressors/biophysical_regressor.py:59]: in GitHub I also see f_sample as an extra tab forward instead of left-aligned with everything else

Comment (karalets) [reply pinot/regressors/biophysical_regressor.py]: check it all now

**Comment by miretchin** [comment]
Pending @yuanqing-wang's comments, it looks good to me now. :)

File: pinot/regressors/biophysical_regressor.py (hunk 8 above)
Comment (karalets) [reply pinot/regressors/biophysical_regressor.py]: oh man, my text editor shows the code formatted very differently than here. 
I'll check its indent options

**Comment by karalets** [comment]
> Pending @yuanqing-wang's comments, it looks good to me now. :)

well this is just the first bit, there is more to come, but thanks for looking it over.

**Comment by yuanqing-wang** [comment]
> Pending @yuanqing-wang's comments, it looks good to me now. :)

I get the idea and I like the overall design a lot. I'll reserve my nitpicking to later.

Also @karalets should I just add a thing for IC50 here? After our discussion it seems not difficult.

**Comment by yuanqing-wang** [comment]
And on that note I think we might use the actual titration curve rather than the Cheng-Prusoff equation.

**Comment by jchodera** [comment]
> And on that note I think we might use the actual titration curve rather than the Cheng-Prusoff equation.

If you need the true multipoint titration data, I can extract if for you from the CDD Vault! Just let me know.

**Comment by karalets** [comment]
I will ping you on the weekend over this, I have a first design for IC50,
what I find more relevant now is the dataset flow etc. to utilize this
stuff correctly.

**Comment by yuanqing-wang** [comment]
> If you need the true multipoint titration data, I can extract if for you from the CDD Vault! Just let me know.
@jchodera it would be very helpful if you could give us the raw data for IC50 curves.

**Comment by karalets** [comment]
I will merge this for now and reopen a PR with more details in this direction.
